File: 3741-minimum-distance-between-three-equal-elements-ii/3741-minimum-distance-between-three-equal-elements-ii.py
from collections import defaultdict
class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        hm = defaultdict(list)
        min_ = float('inf')
        for i in range(len(nums)):
            hm[nums[i]].append(i)

        for candidate in list(hm.values()):
            if len(candidate) < 3: continue
            for m in range(len(candidate)-2):
                i = candidate[m]
                j = candidate[m+1]
                k = candidate[m+2]
                min_ = min(min_, abs(i - j) + abs(j - k) + abs(k - i))
        if min_ == float('inf'):
            return -1 
        return min_

        

File: 3741-minimum-distance-between-three-equal-elements-ii/3741-minimum-distance-between-three-equal-elements-ii.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <cmath>
#include <climits>

using namespace std;

class Solution {
public:
    int minimumDistance(vector<int>& nums) {
        unordered_map<int, vector<int>> hm;
        int min_ = INT_MAX;
        for (int i = 0; i < nums.size(); ++i) {
            hm[nums[i]].push_back(i);
        }
        
        for (const auto& pair : hm) {
            const vector<int>& candidate = pair.second;
            
            if (candidate.size() < 3) continue;
            
            for (size_t m = 0; m < candidate.size() - 2; ++m) {
                int i = candidate[m];
                int j = candidate[m+1];
                int k = candidate[m+2];
                
                int current_dist = abs(i - j) + abs(j - k) + abs(k - i);
                min_ = min(min_, current_dist);
            }
        }
        
        if (min_ == INT_MAX) {
            return -1;
        }
        return min_;
    }
};